### PR TITLE
Client Metrics publisher with KS

### PR DESCRIPTION
**Pull request by kikulikov**
Client Metrics publisher with KS

### CODE REVIEW

File: src/main/avro/Metrics.avsc [hunk 1]
```diff
@@ -0,0 +1,20 @@
+[
```
Comment (astubbs) [inline src/main/avro/Metrics.avsc:1]: can we use the modern avro style?

File: pom.xml [hunk 2]
```diff
@@ -7,11 +7,29 @@
...
+      <url>https://packages.confluent.io/maven/</url>
+    </repository>
+  </repositories>
+
   <dependencies>
     <dependency>
       <groupId>org.apache.kafka</groupId>
       <artifactId>kafka-clients</artifactId>
-      <version>2.2.1</version>
+      <version>2.4.1</version>
```
Comment (astubbs) [inline pom.xml:22]: extract

File: pom.xml [hunk 3]
```diff
@@ -7,11 +7,29 @@
...
+    </dependency>
+    <dependency>
+      <groupId>io.confluent</groupId>
+      <artifactId>kafka-avro-serializer</artifactId>
+      <version>5.4.1</version>
```
Comment (astubbs) [inline pom.xml:27]: extract

File: pom.xml [hunk 4]
```diff
@@ -76,6 +93,23 @@
 
   <build>
     <plugins>
+      <plugin>
+        <groupId>org.apache.avro</groupId>
+        <artifactId>avro-maven-plugin</artifactId>
+        <version>1.9.2</version>
```
Comment (astubbs) [inline pom.xml:99]: extract

File: pom.xml [hunk 5]
```diff
@@ -7,11 +7,29 @@
...
+    </dependency>
+    <dependency>
+      <groupId>io.confluent</groupId>
+      <artifactId>kafka-avro-serializer</artifactId>
+      <version>5.4.1</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.avro</groupId>
+      <artifactId>avro</artifactId>
+      <version>1.9.2</version>
```
Comment (astubbs) [inline pom.xml:32]: extract

File: src/main/java/io/confluent/clientmetrics/avro/MetricNameAVRO.java [hunk 6]
```diff
@@ -0,0 +1,511 @@
+/**
```
Comment (astubbs) [inline src/main/java/io/confluent/clientmetrics/avro/MetricNameAVRO.java:1]: generated -  remove

File: src/main/java/io/confluent/clientmetrics/avro/MetricValueAVRO.java [hunk 7]
```diff
@@ -0,0 +1,310 @@
+/**
```
Comment (astubbs) [inline src/main/java/io/confluent/clientmetrics/avro/MetricValueAVRO.java:1]: remove

**Review by astubbs — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 8]
```diff
@@ -1,88 +1,144 @@
...
-      <version>2.2.1</version>
+      <groupId>org.apache.commons</groupId>
+      <artifactId>commons-lang3</artifactId>
+      <version>${commons-lang3.version}</version>
     </dependency>
     <dependency>
       <groupId>org.testcontainers</groupId>
       <artifactId>kafka</artifactId>
-      <version>1.11.3</version>
+      <version>${testcontainers.version}</version>
```
Comment (astubbs) [inline pom.xml:86]: you only need to extract the version param for the deps that are referenced in multiple places :)

File: src/main/avro/Metrics.avdl [hunk 9]
```diff
@@ -0,0 +1,13 @@
+@namespace("io.confluent.clientmetrics.avro")
```
Comment (astubbs) [inline src/main/avro/Metrics.avdl:1]: muuuch nicer :)

File: src/main/java/io/confluent/clientmetrics/MetricsPublisher.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class MetricsPublisher {
```
Comment (astubbs) [inline src/main/java/io/confluent/clientmetrics/MetricsPublisher.java:28]: I still don't see where this code came from?

**Review by astubbs — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 8 above)
Comment (kikulikov) [reply pom.xml:86]: It looks good and all version numbers in one place )

File: src/main/java/io/confluent/clientmetrics/MetricsPublisher.java (hunk 10 above)
Comment (kikulikov) [reply src/main/java/io/confluent/clientmetrics/MetricsPublisher.java:28]: Done (there were not much code actually :) )

Comment (kikulikov) [reply src/main/java/io/confluent/clientmetrics/MetricsPublisher.java:28]:
```
  private void publishMetrics() {
    mySend(CLIENT_METRICS_TOPIC, "Some metrics...");
    // KStream topology
    // Some JMX metrics
  }
```

File: src/main/java/io/confluent/clientmetrics/MetricsPublisher.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+ * executorService.scheduleAtFixedRate(() -> sendMetrics(producer, streamsClient), 0, 1, TimeUnit.SECONDS);
+ */
+public class MetricsPublisher<V> {
+
+    private static final String DEFAULT_METRICS_TOPIC = "_confluent_client_metrics";
+
+    private final String clientMetricsTopic;
+    private final MetricsSender<V> metricsSender;
+
+    public MetricsPublisher(MetricsSender<V> metricsSender) {
```
Comment (astubbs) [inline src/main/java/io/confluent/clientmetrics/MetricsPublisher.java:29]: implement a default sender

File: src/main/java/io/confluent/clientmetrics/MetricsPublisher.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * Metrics publisher can be scheduled to run periodically, like:
+ * <p>
+ * final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(1);
+ * executorService.scheduleAtFixedRate(() -> sendMetrics(producer, streamsClient), 0, 1, TimeUnit.SECONDS);
```
Comment (astubbs) [inline src/main/java/io/confluent/clientmetrics/MetricsPublisher.java:20]: I dont understand this javadoc?

File: src/test/java/io/confluent/clientmetrics/SchemaRegistryContainer.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
+package io.confluent.clientmetrics;
+
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.KafkaContainer;
+import org.testcontainers.containers.Network;
+
+public class SchemaRegistryContainer extends GenericContainer<SchemaRegistryContainer> {
```
Comment (astubbs) [inline src/test/java/io/confluent/clientmetrics/SchemaRegistryContainer.java:7]: Are you aware of the mock SR we ship?

**Review by astubbs — CHANGES_REQUESTED** [review]
didn't we say you were going to rebase on top of the jenkins branch?

File: src/test/java/io/confluent/clientmetrics/SchemaRegistryContainer.java (hunk 13 above)
Comment (kikulikov) [reply src/test/java/io/confluent/clientmetrics/SchemaRegistryContainer.java:7]: Yes, but it is a pain to configure serdes for the mock SR client.... =(

```java
        final SchemaRegistryClient schemaRegistry = new MockSchemaRegistryClient();

        StringSerializer keySerializer = new StringSerializer(schemaRegistry); // TODO doesn't take the client as a parameter
        SpecificAvroSerializer<AVROMetricRecords> valSerializer = new SpecificAvroSerializer<>(schemaRegistry); // TODO doesn't take the client as a parameter

        final KafkaProducer<String, AVROMetricRecords> producer = new KafkaProducer<>(
                Map.of(
                        ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName(),
                        ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, KafkaAvroSerializer.class.getName(),
                        KafkaAvroSerializerConfig.SCHEMA_REGISTRY_URL_CONFIG, "http://mock:8081",
                        ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka.getBootstrapServers(),
                        ProducerConfig.CLIENT_ID_CONFIG, UUID.randomUUID().toString()
                ), keySerializer, valSerializer);
```

but no such problems with testcontainers.

File: src/main/java/io/confluent/clientmetrics/MetricsPublisher.java (hunk 11 above)
Comment (kikulikov) [reply src/main/java/io/confluent/clientmetrics/MetricsPublisher.java:29]: Why? I do not think I like default hidden assignments. Personally, I would prefer to keep it explicit.

File: src/main/java/io/confluent/clientmetrics/MetricsPublisher.java (hunk 12 above)
Comment (kikulikov) [reply src/main/java/io/confluent/clientmetrics/MetricsPublisher.java:20]: Just an example of how to run it scheduled. Do you want me to remove it?

File: pom.xml [hunk 14]
```diff
@@ -1,113 +1,161 @@
...
+    </repository>
+  </repositories>
+
+  <organization>
+    <name>Confluent</name>
+    <url>https://www.confluent.io/</url>
+  </organization>
+
   <properties>
-    <java.version>1.8</java.version>
```
Comment (astubbs) [inline pom.xml:21]: Put these back in?

File: src/main/java/io/confluent/clientmetrics/MetricsPublisher.java (hunk 12 above)
Comment (astubbs) [reply src/main/java/io/confluent/clientmetrics/MetricsPublisher.java:20]: But the MetricsPublisher isn't mentioned? just the sendMetrics method? which implies it's how it's to be used from within the class. Which given that it's in the class javadoc, breaks the encapsulation of the class.. Is it in the wrong place? Or is it missing something?

File: src/main/java/io/confluent/clientmetrics/MetricsPublisher.java (hunk 11 above)
Comment (astubbs) [reply src/main/java/io/confluent/clientmetrics/MetricsPublisher.java:29]: Just so it makes it easier to use for beginners. Make an opinion, don't be shy.. the user can switch constructors if they want to use their own explicit one.
Sorry, i wasn't clear, i was suggesting to add another constructor which takes no params, which takes no params..

File: src/test/java/io/confluent/clientmetrics/SchemaRegistryContainer.java (hunk 13 above)
Comment (astubbs) [reply src/test/java/io/confluent/clientmetrics/SchemaRegistryContainer.java:7]: yeeeah, but you can reuse that code, and docker is ~slow

File: src/main/java/io/confluent/ps/clientwrapper/ClientWrapper.java [hunk 15]
```diff
@@ -236,83 +229,10 @@ private void buildConsumer(Properties props) {
...
-    });
-    metricKeys.put("appId", getMyId());
-    try {
-      mySend(CLIENT_METRICS_TOPIC, mapper.writeValueAsString(metricKeys));
-    } catch (JsonProcessingException e) {
-      throw new ClientWrapperRuntimeException(e);
-    }
-  }
-
-  private Map getProducerMetricsMap() {
```
Comment (astubbs) [inline src/main/java/io/confluent/ps/clientwrapper/ClientWrapper.java:295]: where was this moved it?

File: src/main/java/io/confluent/ps/clientwrapper/ClientWrapper.java [hunk 16]
```diff
@@ -236,83 +229,10 @@ private void buildConsumer(Properties props) {
...
-    publishJMXMetrics();
-  }
-
-  private void publishJMXMetrics() {
-    Map metrics = wrappedProducer.getDelegateProducer().metrics();
-
-    Set<Entry<MetricName, KafkaMetric>> set = metrics.entrySet();
-
-    // TODO surely there's a better solution than Olog(n) :/ - Hook into Yammer Metrics system?
-    set.iterator().forEachRemaining((e) -> {
```
Comment (astubbs) [inline src/main/java/io/confluent/ps/clientwrapper/ClientWrapper.java:277]: where did this move it?

File: src/main/java/io/confluent/clientmetrics/MetricsPublisher.java (hunk 12 above)
Comment (astubbs) [reply src/main/java/io/confluent/clientmetrics/MetricsPublisher.java:20]: Does this not run periodically by itself anymore?

**Review by astubbs — CHANGES_REQUESTED** [review]

File: src/test/java/io/confluent/clientmetrics/SchemaRegistryContainer.java [hunk 17]
```diff
@@ -0,0 +1,29 @@
+package io.confluent.clientmetrics;
+
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.KafkaContainer;
+import org.testcontainers.containers.Network;
+
```
Comment (astubbs) [inline src/test/java/io/confluent/clientmetrics/SchemaRegistryContainer.java:6]: link to viktor's project

**Pull request closed without merging**